File: 백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.py
import sys
input = sys.stdin.readline
from collections import deque

n, k = map(int, input().split())
a = deque(map(int, input().split()))  # 내구도
robot = deque([0]*n) # n까지만 로봇들이 올라갈 수 있으므로 n개만 생성
cnt = 0

# 내구도가 0인 칸의 개수가 k개 이상이면 종료
def isFinishsed():
  return a.count(0) >= k

while True:
  # 종료 조건
  if isFinishsed():
    break
  
  # 1. 벨트, 로봇 회전하고 내리기
  a.rotate(1)
  robot.rotate(1)
  robot[-1] = 0
  
  # 2. 이동
  if 1 in robot:  # 로봇이 존재 한다면
    for i in range(n-2, -1, -1):  # n과 가까운 것이 먼저 올라간 것
      if robot[i] == 1 and robot[i+1] == 0: # 현재 로봇이 존재하고, 다음 칸에 로봇이 없다면
        if a[i+1] >= 1:  # 만약 내구도가 1 이상 이라면 
          robot[i+1] = 1  # 다음칸으로 이동
          robot[i] = 0  # 현재 칸은 비움
          a[i+1] -= 1
    # 로봇 내리는 곳에 있다면 내리기
    robot[-1] = 0
  
  # 3. 로봇 올리기
  if robot[0] == 0 and a[0] > 0:
    robot[0] = 1  # 로봇 올리기
    a[0] -= 1 # 내구도 -1
  cnt += 1

print(cnt)

File: 백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
	static int n, k, cnt = 0;
	static int[] a, robot;
	
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        
        a = new int[2*n];
        robot = new int[n];
        
        st = new StringTokenizer(br.readLine());
        for (int i=0; i<2*n; i++) {
        	a[i] = Integer.parseInt(st.nextToken());
        }
        
        move();
    }
    
    private static boolean isFinished() {
    	int count = 0;
    	for (int i=0; i<a.length; i++) {
    		if (a[i] == 0) {
    			count++;
    		}
    	}
    	
    	return count >= k;
    }
    
    private static void rotate(int[] arr) {
    	int temp = arr[arr.length-1];
    	for (int i=arr.length-1; i>0; i--) {
    		arr[i] = arr[i-1];
    	}
    	arr[0] = temp;
    }
    
    private static boolean isRobot() {
    	int count = 0;
    	for (int i=0; i<n; i++) {
    		if (robot[i] == 1) {
    			count++;
    		}
    	}
    	
    	return count > 0;
    }
    
    private static void move() {
    	while (true) {
    		// 종료 조건 : 내구도가 0인 칸 k개 이상
    		if (isFinished()) {
    			System.out.print(cnt);
    			break;
    		}
    		
    		// 벨트, 로봇 회전
    		rotate(a);
    		rotate(robot);
    		
        	// 로봇 이동
    		robot[n-1] = 0;	// 로봇 내리기
    		
    		if (isRobot()){	// 내리는 위치 제외
    			for (int i=n-2; i>=0; i--) {
    				// 로봇이 존재하고, 다음 칸에 이동할 수 있으면 (내구도, 로봇 존재)
    				if (robot[i] == 1 && robot[i+1] == 0 && a[i+1] > 0) {
    					robot[i+1] = 1;
    					robot[i] = 0;
    					a[i+1]--;
    				}
    			}
    			robot[n-1] = 0;	// 로봇 내리기
    		}
    		
        	// 시작점 로봇 올리기
    		if (robot[0] == 0 && a[0] > 0) {
    			robot[0] = 1;
    			a[0]--;
    		}
    		
    		cnt++;
    	}
    	
    }
}